internal/api/services/task_service.go

```diff
@@ -0,0 +1,632 @@
+	err = os.MkdirAll(outputDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create submission or output directory: %v", err)
+	}
+
+	// Save the user's file in the submission directory with the correct extension
+	userFilePath := filepath.Join(submissionDir, "solution"+fileExtension)
+	if err := os.WriteFile(userFilePath, userFile, 0644); err != nil {
+		return fmt.Errorf("failed to save user file: %v", err)
+	}
+
+	return nil
+}
+
+// StoreUserOutputs saves output files generated by the user's program inside the appropriate output/ folder
+// under the user's specific submission directory, validating format and matching the task's expected output files.
+func (ts *TaskService) StoreUserOutputs(taskID int, userID int, submissionNumber int, outputFiles map[string][]byte) error {
+	// Define paths for the task, user, and specific submission directories
+	taskDir := filepath.Join(ts.taskDirectory, fmt.Sprintf("task%d", taskID))
+	expectedOutputDir := filepath.Join(taskDir, "src", "output")
+	userSubmissionDir := filepath.Join(taskDir, "submissions", fmt.Sprintf("user%d", userID), fmt.Sprintf("submission%d", submissionNumber))
+	outputDir := filepath.Join(userSubmissionDir, "output")
+
+	// Read expected output files from the task's src/output directory
+	expectedFiles, err := os.ReadDir(expectedOutputDir)
+	if err != nil {
+		return fmt.Errorf("failed to read expected output directory for task %d: %v", taskID, err)
+	}
+
+	// Ensure user submission directory exists
+	if _, err := os.Stat(userSubmissionDir); os.IsNotExist(err) {
+		return fmt.Errorf("submission directory does not exist for task %d, user %d, submission %d", taskID, userID, submissionNumber)
+	}
+
+	// Verify if the output directory already has files
+	if _, err := os.Stat(outputDir); err == nil {
+		entries, err := os.ReadDir(outputDir)
+		if err != nil {
+			return fmt.Errorf("failed to read output directory: %v", err)
+		}
+		if len(entries) > 0 {
+			return fmt.Errorf("output directory already contains files for task %d, user %d, submission %d", taskID, userID, submissionNumber)
+		}
+	} else if os.IsNotExist(err) {
+		// Create the output directory if it doesn't exist
+		if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+			return fmt.Errorf("failed to create output directory: %v", err)
+		}
+	} else {
+		return fmt.Errorf("failed to access output directory: %v", err)
+	}
+
+	// If there's only one file named "compile-error.err", save it and return
+	if len(outputFiles) == 1 {
+		for fileName := range outputFiles {
+			if fileName == "compile-error.err" {
+				return ts.tu.SaveCompileErrorFile(outputDir, outputFiles[fileName])
+			}
+		}
+	}
+
+	// Map expected output numbers from the task's output directory
+	expectedOutputCount := 0
+	expectedNumbers := make(map[int]bool)
+	for _, file := range expectedFiles {
+		if matches := regexp.MustCompile(`^(\d+)\.out$`).FindStringSubmatch(file.Name()); matches != nil {
```

Comment:
No need to create a regex every loop iteration. Define it before and save some CPU execution cycles